### PR TITLE
OOP school library: preserve data

### PR DESCRIPTION
**Pull request by Uzair-Manzoor**
## This pull request has been created against the below mentioned new implementations into the project:

- [x] Created serialize_to_json and deserialize_from_json methods to handle JSON serialization.
- [x] Created PersonFactory module to handle person-related operations.
- [x] Created BookFactory and RentalFactory modules to handle book and rental operations.

### CODE REVIEW

File: person_factory.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  def load_people
+    people_data = deserialize_from_json('people.json')
+    people_data.each do |person_info|
+      person = create_person_from_data(person_info)
+      person.instance_variable_set('@id', person_info['data']['id'])
+      people_manager.people << person
+    end
+  end
+end
```
Comment (Abenezer-Tilahun) [inline person_factory.rb:52]: - [x] As stated in the project requirement **On exit from your app save all the data people,  into JSON files.** but when I checked the people data are saved or not in` json` file on the exit of app , your app throw an error as shown with the below screenshot Kindly Fix the issues.
image

File: rental_factory.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def find_book(book_data)
+    books_manager.books.find { |book| book.title == book_data['title'] && book.author == book_data['author'] }
+  end
+
+  def find_person(person_data)
+    people_manager.people.find { |person| person.name == person_data['name'] }
+  end
+end
```
Comment (Abenezer-Tilahun) [inline rental_factory.rb:46]: - [x] As stated in the project requirement **On exit from your app save all the data rentals,  into JSON files.** but when I checked the rentals data are saved or not in` json` file on the exit of app and try to create a rental  , your app throw an error as shown with the below screenshot Kindly Fix the issues.
image

**Review by Abenezer-Tilahun — CHANGES_REQUESTED** [review]
Hi @Uzair-Manzoor @enocol 

Great job for implemented  the some of the Project Requirement  , but there are some changes that need to be implemented in order to proceed to the next project 👍

## Highlights 👏🏻
✔️ No Linter errors
✔️ Correct Git-Flow
✔️ Good commit messages
✔️ Correctly  save all the data of books in JSON files on the exit of the app and On startup of your app load all data from files.

image
image

almost

## Required Changes ♻️
_Check the comments below the review._

- [x] Kindly make sure that "List all peoples " and "Create a rental" method are working as expected when I checked both of the method it throws  an error as shown below with the below screenshot Kindly fix the issues.
image
image




_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person_factory.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    serialize_to_json(people_data, 'people.json')
+  end
+
+  def load_people
+    people_data = deserialize_from_json('people.json')
+    people_data.each do |person_info|
+      person = create_person_from_data(person_info)
+      person.instance_variable_set('@id', person_info['data']['id'])
+      people_manager.people << person
+    end
```
Comment (aamirkhan2478) [inline person_factory.rb:50]: - [ ] As mentioned by the previous :point_up: reviewer as stated in the project requirement `On exit from your app saves all the data people`, into `JSON` files. but when I checked the people data are saved or not in `json` file on the exit of the app so kindly fix this issue :slightly_smiling_face:

File: rental_factory.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        date: rental.date,
+        book: { title: rental.book.title, author: rental.book.author },
+        person: {
+          id: rental.person.id,
+          age: rental.person.age,
+          name: rental.person.name,
+          parent_permission: rental.person.instance_variable_get('@parent_permission')
+        }
+      }
+    end
```
Comment (aamirkhan2478) [inline rental_factory.rb:24]: - [ ] As mentioned by the previous :point_up: reviewer as stated in the project requirement `On exit from your app save all the data rentals`, into `JSON` files. but when I checked the rental data were saved or not in `json` file on the exit of app so kindly fix this issue as well :slightly_smiling_face:

**Review by aamirkhan2478 — CHANGES_REQUESTED** [review]
Hi @Uzair-Manzoor & @enocol, 



Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

 ## Highlights
✔️ Good Effort
✔️ Your design is good.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Uzair-Manzoor** [comment]
Hi @aamirkhan2478, thanks for the feedback. The app is working perfectly with all functionalities after recent amendments. Screenshots are attached for clear understanding

image

image

image

image

image

**Review by omar-muhamad — APPROVED** [review]
# Approved ✔
### Hi @,

#### Great job, Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

### Congratulations! 🎉 

giphy (1)

## (Highlights) Good Points: :+1:
- Everything works perfectly and the data is saved in JSON files. ✔
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- N/A

## [Optional] suggestions:

- N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**